### ISSUE TITLE
Unresponsive editor

### ISSUE THREAD

**Issue by fekga** [markusfisch/ShaderEditor#31]
Any shader that increases in size gets really unresponsive, deleting/adding characters become **really** slow. Even scrolling will be laggy. I'm using it by turning off the compiling on editing but it isn't helping. Please increase responsiveness on the texteditor.

**Comment by markusfisch** [comment]
Unfortunately, it's not ShaderEditor.highlight() nor the regular expressions but rather Android's TextView.

Ran some quick tests and here are the times for 1124 lines of code on a Nexus 4:

clearSpans(): 242ms
PATTERN_NUMBERS.matcher(): 1ms
482x setSpan: 71ms
PATTERN_PREPROCESSOR.matcher(): 1ms
3x setSpan: 6ms
PATTERN_KEYWORDS.matcher(): 1ms
463x setSpan: 99ms
PATTERN_BUILTINS.matcher(): 1ms
384x setSpan: 89ms
PATTERN_COMMENTS.matcher(): 1ms
0x setSpan: 0ms
total: 512ms

So it takes around 512ms to (allocate and) insert *a lot* of new objects. Half of the time is clearing spans. *After* that, the EditText holding all the Spannables is barely usable anymore. Guess I have to dive into android.widget.TextView but I'm afraid there's not much to do. Except from trying to draw the text manually.

**Comment by fekga** [comment]
Just by searching "android textview slow" I found this article about it and
a possible solution:
https://www.wattpad.com/245272242-adventures-in-androidland-slow-textview-is-slow

2016-08-13 18:40 GMT+02:00 Markus Fisch <notifications@github.com>:

**Comment by fekga** [comment]
It would partially solve this problem if you could use multiple shader sources, one or more as a "library" for useful functions and definitions and a much shorter source as your shader program. Each source could then have a window where you could edit them. This way there wouldn't be that many Spannables and would also add a reusable (library-like) source as feature.

**Comment by markusfisch** [comment]
That's be an interesting possibility! But I'm not ready to give up yet :) Plus the necessary UI wouldn't be trivial I suppose.

I'm still examining that solution you posted (good info, btw). The symptoms they have are quite the same but then they only seem to have a delay when inserting new characters. The many spanables required to do syntax highlighting are killing even the scrolling of the text.

**Comment by markusfisch** [comment]
Just a note; Nougat's SpannableStringBuilder seems to perform a lot better. While the initial highlighting takes some time there too, scrolling works quite okay even with 2000+ lines of code.

**Comment by markusfisch** [comment]
Well, there's a setting to disable syntax highlighting for large sources (> 4096 characters) since 2.12.0: https://github.com/markusfisch/ShaderEditor/commit/62751083199600a3a2fcec7fc767af519f2bb82b

But the app still uses the default EditText implementation. Sorry for that.

**Comment by fekga** [comment]
It seems that the running speed of the shader affects the responsiveness of the editor. Lower FPS on the shader causes slower editor update times, even without syntax highlight. It still seems like it parses the whole shader source with every character input and not when running the shader.

**Comment by markusfisch** [comment]
That's right, the running speed of the shader definitely affects the UI, unfortunately. Although I'm not entirely sure why, because all the rendering happens in another thread already. Troubles me for quite some time, but I haven't found the time to really look into it.

About the parsing, yes, the app parses the whole text after each edit to remove (potential) error highlights even with syntax highlighting switched off. But that shouldn't be a problem even with large shaders - at least it wasn't a problem when I did some profiling on low-end devices. So, are you sure it's the parsing or is it maybe the rendering thread blocking the UI thread?

Is editing still too slow if you try a shader that runs really fast but has a lot of (blind) code?
